include/xtensor/xstrides.hpp

```diff
@@ -71,30 +71,57 @@ namespace xt
     inline auto compute_size(const shape_type& shape) noexcept
     {
         using size_type = std::decay_t<typename shape_type::value_type>;
-        return std::accumulate(shape.begin(), shape.end(), size_type(1), std::multiplies<size_type>());
+        return std::accumulate(shape.cbegin(), shape.cend(), size_type(1), std::multiplies<size_type>());
     }
 
-    template <class size_type, class S, size_t dim>
-    inline size_type data_offset(const S& /*strides*/) noexcept
+    namespace detail
+    {
+        template <class size_type, class S, std::size_t dim>
+        inline size_type raw_data_offset(const S&) noexcept
+        {
+            return 0;
+        }
+
+        template <class size_type, class S, std::size_t dim, class Arg, class... Args>
+        inline size_type raw_data_offset(const S& strides, Arg arg, Args... args) noexcept
+        {
+            return arg * strides[dim] + raw_data_offset<size_type, S, dim + 1>(strides, args...);
+        }
+    }
+
+    template <class size_type, class S, std::size_t dim>
+    inline size_type data_offset(const S&) noexcept
     {
         return 0;
     }
 
-    template <class size_type, class S, size_t dim, class... Args>
-    inline size_type data_offset(const S& strides, size_type i, Args... args) noexcept
+    template <class size_type, class S, std::size_t dim, class Arg, class... Args>
+    inline size_type data_offset(const S& strides, Arg arg, Args... args) noexcept
     {
-        if (sizeof...(Args) + 1 > strides.size())
+        constexpr std::size_t nargs = sizeof...(Args) + dim + 1;
+        if (nargs == strides.size())
```

Comment:
In the case of `xtensor`, where `strides.size()` is constexpr, this should be optimized by the compiler.